javalib/src/main/scala/java/io/CharArrayReader.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.io
+
+class CharArrayReader(protected var buf: Array[Char], offset: Int, length: Int) extends Reader {
+  if (offset < 0 || offset > buf.length || length < 0 || offset + length < 0) throw new IllegalArgumentException
+
+  protected var pos: Int = offset
+  protected var markedPos: Int = offset
+  protected var count: Int = {
+    val bufferLength = buf.length
+
+    if (offset + length < bufferLength) length
+    else bufferLength
+  }
+
+  def this(buf: Array[Char]) = {
+    this(buf, 0, buf.length)
+  }
+
+  override def close(): Unit = if (isOpen) this.buf = null
+
+  private def isOpen: Boolean = this.buf != null
+  private def isClosed: Boolean = !isOpen
+
+  override def mark(readAheadLimit: Int): Unit = {
+    ensureOpen()
+
+    // The parameter readLimit is ignored for CharArrayReaders
+    this.markedPos = this.pos
+  }
+
+  override def markSupported(): Boolean = true
+
+  override def read(): Int = {
+    ensureOpen()
+
+    if (pos == count) {
+      -1
+    } else {
+      pos += 1
+      buf(pos - 1)
+    }
+  }
+
+  override def read(buffer: Array[Char], offset: Int, len: Int): Int = {
+    if (offset < 0 || offset > buffer.length)
+      throw new ArrayIndexOutOfBoundsException("Offset out of bounds : " + offset)
+
+    if (len < 0 || len > buffer.length - offset)
+      throw new ArrayIndexOutOfBoundsException("Length out of bounds : " + len)
+
+    ensureOpen()
+
+    if (this.pos < this.count) {
+      val bytesRead = Math.min(len, count - pos)
+      System.arraycopy(this.buf, this.pos, buffer, offset, bytesRead)
+      this.pos += bytesRead
+      bytesRead
+    } else {
+      -1
+    }
+  }
+
+  override def ready(): Boolean = {
+    // JDK says "Character-array readers are always ready to be read.", however,
+    // CharArrayReaderTest.ready_should_throw_IOException passes on the JVM, so
+    // we also throw the exception
```

Comment:
I don't think @gzm0's previous comment (https://github.com/scala-js/scala-js/pull/4162#discussion_r482041607) was about throwing the exception, but about returning `false` when `pos == count`. I think that throwing an exception when the reader is closed is 100% by-spec, as all methods are supposed to throw once it's closed.

What's not by-spec is to return `false` when `pos == count`. Per spec, even that should return `true`, because we would not *block waiting for input* (we would immediately return `-1` in `read()`).

Reply:
Thanks for the explanation, I was obviously not focused on the right issue.  I just verified OpenJDK8 does return false (it is covered by the existing test cases), should I match the JVM or match the spec?

Reply:
Always match the JVM, but if that deviates from the spec, it's good to call it out in a comment.